feat(file-access-mode-constants): introduce

nekojanai opened the pull request:
closes https://github.com/nodejs/userland-migrations/issues/130

WIP

AugustinMauroy commented on recipes/file-access-mode-constants/package.json:
```diff
@@ -0,0 +1,25 @@
+  "author": "nekojanai (Jana)",
+  "license": "MIT",
+  "homepage": "https://github.com/nodejs/userland-migrations/blob/main/recipes/file-access-mode-constants/README.md",
+  "dependencies": {
+    "@ast-grep/napi": "^0.39.2",
+    "@nodejs/codemod-utils": "^0.0.0"
+  },
```
```suggestion
  "dependencies": {
    "@nodejs/codemod-utils": "*"
  },
```

`*` mean monorepo one

AugustinMauroy commented on recipes/fs-access-mode-constants/src/workflow.ts:
```diff
@@ -0,0 +1,283 @@
+import type { Edit, SgRoot } from "@codemod.com/jssg-types/main";
+import { getNodeRequireCalls } from "@nodejs/codemod-utils/ast-grep/require-call";
```
```suggestion
port type Js from "@codemod.com/jssg-types/langs/javascript";
import { getNodeRequireCalls } from "@nodejs/codemod-utils/ast-grep/require-call";
```

You can import the grammar to have a better typecking in query and usage is `SgNode<Js>` or `SgRoot<Js>`

JakobJingleheimer commented on recipes/fs-access-mode-constants/src/workflow.ts:
```diff
@@ -0,0 +1,284 @@
+import { getNodeImportStatements } from "@nodejs/codemod-utils/ast-grep/import-statement";
+import type Js from "@codemod.com/jssg-types/langs/javascript";
+
+export default function tranform(root: SgRoot<Js>): string | null {
```
Is there perhaps a type for this function you can apply via annotation, like `const transform: Workflow = (root) => {`?

JakobJingleheimer commented on recipes/fs-access-mode-constants/src/workflow.ts:
```diff
@@ -0,0 +1,284 @@
+  let hasPromisesImport = false;
+  let promisesImportName = "promises";
+
+  // @ts-ignore - ast-grep types are not fully compatible with JSSG types
```
I think better to use @ts-expect-error

JakobJingleheimer commented on recipes/fs-access-mode-constants/src/workflow.ts:
```diff
@@ -0,0 +1,284 @@
+        originalText = newText;
+      }
+
+      if (originalText.includes("X_OK")) {
```
Can these be combined? I think that will help readers to understand what is happening.

nekojanai replied:
I'm not sure I know exactly how to combine those two lines to make it more expressive...
...suggestions?

AugustinMauroy commented on package-lock.json:
what Claudio ?

nekojanai replied:
> It looks like the root level ast-grep dep has been updated, which should happen in a separate PR.
> 
> PS Sorry if I'm mistaken; pjson lock is quite difficult to read (yes, @ovflowd, you were right 😅)

Yeah, I updated that package and then reverted the update but the changes to the package-lock.json stuck, removing them for good now, good catch!

ovflowd replied:
🍿

brunocroh commented on recipes/fs-access-mode-constants/src/workflow.ts:
```diff
@@ -0,0 +1,284 @@
+        originalText = newText;
+      }
+
+      if (originalText.includes(", }") || originalText.includes(",  ") || originalText.includes(" , ")) {
+        const newText = originalText.replace(" , ", "").replace(",,", ",").replace(",  ", ", ").replace(", }", " }");
```
Yes, we have an utility function `utils/src/ast-grep/remove-binding.ts` that can be used here

```js
  let hasChange = false
  for (const key of ['F_OK', 'R_OK', 'W_OK', 'X_OK']) {
    const change = removeBinding(importStatement, key); //it will return a ast-grep.Edit to remove import if it exists
    if(change) {
       edits.push(change.edit)
       hasChange = true
    }
  }
  
  if(hasChange) {
    // add constants to import
  }
```

brunocroh replied:
An easier way to handle destructured objects or named imports is

```js
// get items inside of objectPattern
const imports = objectPattern.find({
  rule: {
    any: [
     {kind: 'import_specifier'},
     {kind: 'shorthand_property_identifier_pattern'}
    ]
  }
})
//now imported items will be in `imports` variable as a string array, that will be easier to transform, i.g
// imports.push('newImport')
// items.filter // remove import
// items.map // update import

// after transform just join the array with ", "
const newImportString = `{ ${newImports.join(", ")} }`

// Replace the entire object pattern with new object pattern
objectPattern.replace(newImportString)

```

nekojanai replied:
> An easier way to handle destructured objects or named imports is
> 
> ```js
> // get items inside of objectPattern
> const imports = objectPattern.find({
>   rule: {
>     any: [
>      {kind: 'import_specifier'},
>      {kind: 'shorthand_property_identifier_pattern'}
>     ]
>   }
> })
> //now imported items will be in `imports` variable as a string array, that will be easier to transform, i.g
> // imports.push('newImport')
> // items.filter // remove import
> // items.map // update import
> 
> // after transform just join the array with ", "
> const newImportString = `{ ${newImports.join(", ")} }`
> 
> // Replace the entire object pattern with new object pattern
> objectPattern.replace(newImportString)
> ```

I tried getting the import specifiers in an array via the ruleset provided above but going off of
```ts
    const objectPattern = statement.find({
      rule:
        { kind: 'object_pattern' },
    });
```
that didn't yield any results so I took a different approach

nekojanai commented:
incorporated feedback and adjusted coded accordingly
@brunocroh I choose to not use the utility function mentioned

my current approach allows me to handle every require and import statements with arrays containing the specifiers, and conditional replacements instead of removal is more streamlined imo

brunocroh commented:
> incorporated feedback and adjusted coded accordingly @brunocroh I choose not to use the utility function mentioned
> 
> my current approach allows me to handle every require and import statements with arrays containing the specifiers, and conditional replacements instead of removal is more streamlined imo

That's fine, don't worry!

JakobJingleheimer commented on recipes/fs-access-mode-constants/src/workflow.ts:
```diff
@@ -0,0 +1,71 @@
+    });
+
+    if (objectPattern) {
+      let objPatArr = objectPattern.text().split('{ ')[1].split(' }')[0].split(', ');
```
I think this is still outstanding?

PS If you can't go the AST route, perhaps regex could be robust enough here: `/{\s+/` instead of `'{ '` (I think I had suggested some regex alternatives before).

nekojanai replied:
@brunocroh that `.find` call doesn't produce an iterable structure
I'm not sure what I'm doing wrong here but the suggested change does not work.

JakobJingleheimer reviewed (commented):
I think this is very close!

P.S. Sorry for my lag (been very busy).

JakobJingleheimer commented:
@nekojanai **please** stop force-pushing. Doing that makes it very difficult to review changes over time; instead of being able to see the subset of new changes, the reviewer has look at _everything_ again because the periodic diff is broken.
